### PR TITLE
Extend the color picker to change between hex, rgb, rgba

### PR DESCRIPTION
**Pull request by jjaracz**
Pull Request for Issue #14353 .

### Summary of Changes

Adds buttons to change between hex, rgb, rgba.

### Testing Instructions

Add to protostar templateDetails.xml code below to the xml `extension -> config -> fields` and go to settings of that template and then to the advanced tab. There should be new extended color pickers.

```xml
<field name="Color1" type="color" label="Color + changeformat" changeformat="true" />
<field name="Color2" type="color" label="Color + keywords" format="hex" keywords="transparent,inherit,initial" />
<field name="Color3" type="color" label="Color + format" format="rgb" />
<field name="Color4" type="color" label="Color" />
```
`keywords` - if you add this attribute to field then you can type to the input keywords that you specify and save it.
`format` - you can choose color only from specified format of color.
`changeformat` - it enables you to save color in 3 formats `rgb`, `hex` and `rgba`

### Documentation Changes Required

Yes on the page [https://docs.joomla.org/Color_form_field_type](https://docs.joomla.org/Color_form_field_type)

- boolean attribute `changeformat=""` is to show buttons to change format between hex, rgb, rgba.
- format and keywords attribute from #10129

zaznaczenie_134

### CODE REVIEW

**Comment by franz-wohlkoenig** [comment]
I have tested this item :white_check_mark: successfully on 5e56bc4e1c4a2ce21de4e41865d9f233b049d859<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14427">issues.joomla.org/tracker/joomla-cms/14427</a>.</sub>

**Comment by Bakual** [comment]
Are we sure the other formats work in all cases where that field is used?
Otherwise it needs to be opt in based on attributes in the XML definition.

**Comment by ciar4n** [comment]
There is an existing option with the color field to set to an RGBA format... https://github.com/joomla/joomla-cms/pull/10129. How would this PR effect that?

**Comment by jjaracz** [comment]
@ciar4n I resolve problem in different way because author of that pr you mention, adds atr `format=""` in xml, but i recognize it from value and adds buttons to dynamic change between formats.

@Bakual I think It isn't necessary, because i recognize the format from value. Unless you mean to specify one format without change to others ?

File: media/system/css/fields/color.css [hunk 1]
```diff
@@ -0,0 +1,28 @@
+.minicolors.minicolors-input{
```
Comment (C-Lodder) [inline media/system/css/fields/color.css]: Should be a space before each `{`

`.foo{` >> `.foo {`

https://developer.joomla.org/coding-standards/css.html

File: libraries/joomla/form/rule/color.php [hunk 2]
```diff
@@ -43,16 +43,25 @@ public function test(SimpleXMLElement $element, $value, $group = null, Registry
 			return false;
 		}
 
-		if ($value[0] != '#')
+
+		if($value[0] == "#")
```
Comment (C-Lodder) [inline libraries/joomla/form/rule/color.php]: Spacing:

`if(` >> `if (`

https://developer.joomla.org/coding-standards/php-code.html

File: layouts/joomla/form/field/color/advanced.php [hunk 3]
```diff
@@ -45,20 +45,20 @@
  * @var   array    $control         Is this field checked?
  */
 
+$format       = $color[0] != '#' && strpos($color, 'rgb') === 0 ? (strpos($color, 'rgba') === 0 ? 'rgba' : 'rgb') : 'hex';
 if (in_array($format, array('rgb', 'rgba')) && $validate != 'color')
 {
-	$alpha = ($format === 'rgba');
-	$placeholder = $alpha ? 'rgba(0, 0, 0, 0.5)' : 'rgb(0, 0, 0)';
+		$alpha = ($format === 'rgba');
```
Comment (C-Lodder) [inline layouts/joomla/form/field/color/advanced.php]: only need 1 tab indentation, not 2

File: layouts/joomla/form/field/color/advanced.php [hunk 4]
```diff
@@ -45,20 +45,20 @@
...
+		$placeholder = $alpha ? 'rgba(0, 0, 0, 0.5)' : 'rgb(0, 0, 0)';
 }
 else
 {
-	$placeholder = '#rrggbb';
+		$placeholder = '#rrggbb';
```
Comment (C-Lodder) [inline layouts/joomla/form/field/color/advanced.php]: same here ^^

File: media/system/js/color-field-adv-init.js [hunk 5]
```diff
@@ -1,17 +1,69 @@
-jQuery(document).ready(function ($){
-	$('.minicolors').each(function() {
-		$(this).minicolors({
+jQuery(document).ready(function ($) {
+	// Array of rendered color-pickers
+	var $miniColors = [],
+		lineNum, $input, format;
+
+
```
Comment (yvesh) [inline media/system/js/color-field-adv-init.js]: Remove one empty line

File: layouts/joomla/form/field/color/advanced.php [hunk 6]
```diff
@@ -72,11 +75,21 @@
...
-$position; ?><?php echo $control; ?><?php echo $readonly; ?><?php echo $disabled; ?><?php echo
-$required; ?><?php echo $onchange; ?><?php echo $autocomplete; ?><?php echo $autofocus; ?><?php echo
-$format; ?><?php echo $keywords; ?><?php echo $direction; ?><?php echo $validate; ?>/>
+<div class="color-picker-group">
+	<input type="text" name="<?php echo $name; ?>" id="<?php echo $id; ?>" value="<?php
+	echo htmlspecialchars($color, ENT_COMPAT, 'UTF-8'); ?>"<?php echo $hint; ?><?php echo $class; ?><?php echo
+	$position; ?><?php echo $control; ?><?php echo $readonly; ?><?php echo $disabled; ?><?php echo
+	$required; ?><?php echo $onchange; ?><?php echo $autocomplete; ?><?php echo $autofocus; ?><?php echo
+	$dataformat; ?><?php echo $keywords; ?><?php echo $direction; ?><?php echo $validate; ?>/>
+	<?php if($showbtns): ?>
```
Comment (C-Lodder) [inline layouts/joomla/form/field/color/advanced.php]: `<?php if($showbtns): ?>` >> `<?php if ($showbtns) : ?>`

File: layouts/joomla/form/field/color/advanced.php [hunk 7]
```diff
@@ -72,11 +75,21 @@
...
+		<div class="color-format-group">
+			<button class="color-format-btn <?php echo $format == 'hex' ? 'active' : '' ?>" type="button" data-format="hex">HEX</button>
+			<button class="color-format-btn <?php echo $format == 'rgb' ? 'active' : '' ?>" type="button" data-format="rgb">RGB</button>
+			<button class="color-format-btn <?php echo $format == 'rgba' ? 'active' : '' ?>" type="button" data-format="rgba">RGBA</button>
+		</div>
+	<?php endif;?>
```
Comment (C-Lodder) [inline layouts/joomla/form/field/color/advanced.php]: spacing: 

`<?php endif;?>` >> `<?php endif; ?>`

**Comment by N6REJ** [comment]
the reason for the format="" was to allow for transparency!  It's important to have the ability to have color: transparent"  look @ previous pr's regarding this issue.

**Comment by N6REJ** [comment]
@brianteeman something seriously wrong with this pr.. when I look @ the first 2 changes in the commit they are about files that are not part of this pr.

**Comment by jjaracz** [comment]
@N6REJ attributes keywords and format are still working :)

**Comment by jjaracz** [comment]
@C-Lodder @yvesh Thank you :slightly_smiling_face:

**Comment by Bakual** [comment]
> I think It isn't necessary, because i recognize the format from value. Unless you mean to specify one format without change to others ?

@jjaracz The reason I ask isn't about the part where the JS reads the value. I fully expect that to be able to deal with the various formats. My concern is the other direction, the one receiving the value set. If the extension expects a HEX value and you pass a RGB one, it may break the extension.
I think you will have to default to HEX only and the extension developer can specify additional supported formats in the XML definition.

**Comment by brianteeman** [comment]
@N6REJ no point tagging me I can't do anything

**Comment by jjaracz** [comment]
@Bakual You're right, I will try to do this :slightly_smiling_face:

**Comment by N6REJ** [comment]
@jjaracz should there be !!$color? https://github.com/joomla/joomla-cms/pull/14427/files#diff-c84005990b5f00eea2cc54b543c1fb22R219  Isn't it just !$color ?

**Comment by jjaracz** [comment]
@N6REJ No, because further I can't check first letter in a blank string so I should eliminate it :slightly_smiling_face:

**Comment by laoneo** [comment]
Can you fix the merge conflicts please, then we should do another round of testing.

**Comment by C-Lodder** [comment]
Will also need to be tested on sub form fields. Not sure it will work as you haven't delegated the click event

**Comment by jjaracz** [comment]
@laoneo 
@C-Lodder 
Now should be ok :)

**Comment by franz-wohlkoenig** [comment]
@jjaracz can you please update your Test Instructions and first Comment like `changeformat="true"` is now in templateDetails.xml and Screen show:
bildschirmfoto 2017-06-15 um 16 03 57

File: layouts/joomla/form/field/color/advanced.php [hunk 8]
```diff
@@ -73,8 +77,21 @@
...
-echo $hint . $class . $position . $control . $readonly . $disabled . $required . $onchange . $autocomplete . $autofocus . $format . $keywords . $direction . $validate; ?>/>
+<div class="color-picker-group">
+	<input type="text" name="<?php echo $name; ?>" id="<?php echo $id; ?>" value="<?php
+	echo htmlspecialchars($color, ENT_COMPAT, 'UTF-8'); ?>"<?php echo $hint . $class .
+	$position . $control . $readonly . $disabled .
+	$required . $onchange . $autocomplete . $autofocus .
+	$dataformat . $keywords . $direction . $validate; ?>/>
+	<?php if ($changeFormat) : ?>
+		<div class="color-format-group">
+			<button class="color-format-btn <?php echo ($format == 'hex' ? 'active' : ''); ?>" type="button" data-format="hex">HEX</button>
```
Comment (Quy) [inline layouts/joomla/form/field/color/advanced.php]: I assume `HEX`, `RGB`, and `RGBA` should be language strings.

Comment (laoneo) [inline layouts/joomla/form/field/color/advanced.php]: True.

Comment (jjaracz) [inline layouts/joomla/form/field/color/advanced.php]: @Quy
@laoneo 
Do you know what name has language file for that path? I couldn't find :/

**Comment by jjaracz** [comment]
@franz-wohlkoenig 
updated ;)

File: layouts/joomla/form/field/color/advanced.php (hunk 8 above)
Comment (laoneo) [inline layouts/joomla/form/field/color/advanced.php]: there is one on the front as well one on the back end, @infograf768 can probably answer that question better

**Comment by franz-wohlkoenig** [comment]
@jjaracz didn't get Fields like you show in first Comment. What i get is:
27185109-9693622c-51e4-11e7-8dc8-391ac55fb009
by adding `changeformat="true` in Protostar `templateDetails.xml` in field of type `colors`.

**Comment by jjaracz** [comment]
@franz-wohlkoenig 
```xml
<field type="color" ... keywords="transparent,inherit,initial" />
<field type="color" ... format="rgb" />
```
This should work :)

**Comment by franz-wohlkoenig** [comment]
@jjaracz can you please update your Test Instructions so willing Testers without Dev.-Knowledge can test easy?

**Comment by jjaracz** [comment]
@franz-wohlkoenig
Sorry for that, I was thinking that was obvious :)

**Comment by franz-wohlkoenig** [comment]
@jjaracz thanks for Info. 

Copied code in `templateDetails.xml`:
```<fields name="params">          
			<fieldset name="advanced">
              <field name="Color1" type="color" label="Color + changeformat" changeformat="true" />
<field name="Color2" type="color" label="Color + keywords" format="hex" keywords="transparent,inherit,initial" />
<field name="Color3" type="color" label="Color + format" format="rgb" />
<field name="Color4" type="color" label="Color" />
				<field 
					name="templateColor" 
					type="color" 
					label="TPL_PROTOSTAR_COLOR_LABEL"
					description="TPL_PROTOSTAR_COLOR_DESC" 
					class="" 
					default="#08C"
				/>
(...)
		</fields>
``` 

got:
bildschirmfoto 2017-06-19 um 11 12 56
So "Color + changeformat" looks different.

**Comment by franz-wohlkoenig** [comment]
@jjaracz thanks for Info. Using PR and add Color to the new Fields doesn't change Frontend – expected?

**Comment by ghost** [comment]
If this Issue get no Response, it will be closed at 17th September 2017.

**Comment by laoneo** [comment]
Why you want to close it, this is a useful feature, it just lacks of testers.

**Comment by ghost** [comment]
As there was no Response for 2 month.

**Comment by ghost** [comment]
This has been closed due to lack of response to the requests above – it can always be reopened.

**Comment by ghost** [comment]
Reopened as #14353 needs this PR to be merged.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14427">issues.joomla.org/tracker/joomla-cms/14427</a>.</sub>

**Comment by roland-d** [comment]
@jjaracz Could you fix the conflicts please?

**Comment by cyrez** [comment]
@roland-d Maybe this PR is outdated due to new color picker planned for j4 : https://github.com/joomla/joomla-cms/issues/14353#issuecomment-330025770

**Comment by roland-d** [comment]
@JoomliC I don't think this is outdated. This is the only PR that exists for extending the color picker, so if this can get merged it would be great. Any future changes can be done based on this code.

**Comment by priiish** [comment]
I have tested this item :white_check_mark: successfully on cc572e1f52b1b51a9fedc669b1a2a5f7e683015d<br /><br /># Testing Steps

- Opened templateDetails.xml, inserted code snippet, went to Extensions/Templates/Templates/Styles/Protostar/Advanced Settings
- Tested color picker -> format change works fine, data is being saved correctly into database (Joomla; table: ibloj_template_styles)
- see screenshots below for datailed documentation

@icampus

<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14427">issues.joomla.org/tracker/joomla-cms/14427</a>.</sub>

**Comment by priiish** [comment]
Screenshots regarding test on Issue #14427
tested successfully 
-> changed color formats being saved correctly into database
bildschirmfoto 2018-07-23 um 12 52 00

bildschirmfoto 2018-07-23 um 12 49 29

**Comment by kneisel** [comment]
I have tested this item :red_circle: unsuccessfully on cc572e1f52b1b51a9fedc669b1a2a5f7e683015d<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14427">issues.joomla.org/tracker/joomla-cms/14427</a>.</sub>

**Comment by kneisel** [comment]
Works "partly".
"Good cases" are working, but other cases can be constructed that work unexpectedly

case: 
field name="Color1" type="color" label="Color + changeformat" changeformat="true"
format is set to RGBA but color is displayed as HEX
image

case:
field name="Color3" type="color" label="Color + format" format="rgba"
format is set to "rgba" but it is possible to write hex
image

**Comment by Schmidie64** [comment]
I have tested this item :red_circle: unsuccessfully on cc572e1f52b1b51a9fedc669b1a2a5f7e683015d<br /><br />I followed the instructions step by step. In some cases the patch works fine for example by changing the Template Color. At the Field were i can pick a format and pick a color, the colorpicker takes in every situtation the Hex-color-code. And when i pick a color and for example rgba is checked, it put in the hex-color-code. But when I switch to the hex button, i've got an invalid hex code. 

image

So somewere in the code must be a conversion error.

@icampus <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14427">issues.joomla.org/tracker/joomla-cms/14427</a>.</sub>

**Comment by kofaysi** [comment]
I have tested this item :red_circle: unsuccessfully on cc572e1f52b1b51a9fedc669b1a2a5f7e683015d<br /><br />J3.9.2

Going directly to the "Template colour" and cycling through the options (hex, rgb, rgba), the value #0088cc has been reset to rgba(0, 0, 0, 1) and does not change back, when applying rgb or hex.

The "Background Colour" colour selector does not have hex, rgb or rgba variants available.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14427">issues.joomla.org/tracker/joomla-cms/14427</a>.</sub>

**Comment by Quy** [comment]
@brianteeman OK to close this?

**Comment by HLeithner** [comment]
@jjaracz thx for your time you have invested into this PR but J3 there will be no new features.
For J4 I'm not sure whats the color picker is but i'm also not sure if this is a feature a user really needs.
It's maybe more confusing an end user if s/he could switch between formats without knowing what this nice values mean.

Anyway I'm closing this for J3, if you like you can look into J4 and maybe can help making the j4 color picker better (or anything else you like to help with).

thx for your investment in Joomla!

**Comment by N6REJ** [comment]
sad that this took 2 yrs just to NOT be implemented.

**Pull request closed without merging**